pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/relocation/RealtimeSegmentRelocationManager.java

```diff
@@ -0,0 +1,287 @@
+
+
+/**
+ * Manager to relocate completed segments from consuming servers to completed servers
+ * Segment relocation will be done by this manager, instead of directly moving segments to completed servers on completion,
+ * so that we don't get segment downtime when a segment is in transition
+ *
+ * We only relocate segments for realtime tables, and only if tenant config indicates that relocation is required
+ * A segment will be relocated, one replica at a time, once all of its replicas are in ONLINE state and on consuming servers
+ */
+public class RealtimeSegmentRelocationManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(RealtimeSegmentRelocationManager.class);
+  private final ScheduledExecutorService _executorService;
+  private final PinotHelixResourceManager _pinotHelixResourceManager;
+  private final HelixManager _helixManager;
+  private final HelixAdmin _helixAdmin;
+  private final long _runFrequencyMinutes;
+
+  public RealtimeSegmentRelocationManager(PinotHelixResourceManager pinotHelixResourceManager, ControllerConf config) {
+    _pinotHelixResourceManager = pinotHelixResourceManager;
+    _helixManager = pinotHelixResourceManager.getHelixZkManager();
+    _helixAdmin = pinotHelixResourceManager.getHelixAdmin();
+    _runFrequencyMinutes = config.getRelocationManagerFrequencyInMinutes();
+
+    _executorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
+      @Override
+      public Thread newThread(Runnable runnable) {
+        Thread thread = new Thread(runnable);
+        thread.setName("PinotRelocationManagerExecutorService");
+        return thread;
+      }
+    });
+  }
+
+  public void start() {
+    LOGGER.info("Starting relocation manager");
+
+    _executorService.scheduleWithFixedDelay(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          runRelocation();
+        } catch (Exception e) {
+          LOGGER.warn("Caught exception while running relocation manager", e);
+        }
+      }
+    }, 5, _runFrequencyMinutes, TimeUnit.MINUTES);
+  }
+
+  public void stop() {
+    _executorService.shutdown();
+  }
+
+  /**
+   * Check all tables. Perform relocation of segments if table is realtime and relocation is required
+   */
+  public void runRelocation() {
+    if (!_pinotHelixResourceManager.isLeader()) {
+      LOGGER.info("Skipping realtime segment relocation, not leader!");
+      return;
+    }
+
+    LOGGER.info("Starting relocation of realtime segments");
+    List<String> allTableNames = _pinotHelixResourceManager.getAllTables();
+
+    for (String tableNameWithType : allTableNames) {
+
+      TableType tableType = TableNameBuilder.getTableTypeFromTableName(tableNameWithType);
+      if (tableType.equals(TableType.REALTIME)) {
+        LOGGER.info("Starting relocation of segments for table: {}", tableNameWithType);
+
+        TableConfig tableConfig = _pinotHelixResourceManager.getRealtimeTableConfig(tableNameWithType);
```

Comment:
Do the whole thing under a try/catch, moving on (after logging the table name that had a problem) if there is an issue with one table. Otherwise, we may never move past a table that had any exception for whatever reason.

Reply:
Ah yes, good suggestion